libraries/node-core-library/src/FileError.ts

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+import * as NodeJSPath from 'path';
+
+import { Path } from './Path';
+import { TypeUuid } from './TypeUuid';
+
+/**
+ * The format that the FileError message should conform to. The supported formats are:
+ *  - Unix: \<filePath\>:\<line\>:\<column\> - \<message\>
+ *  - VisualStudio: \<filePath\>(\<line\>,\<column\>) - \<message\>
+ *  - AzureDevOps: ##vso[task.logissue type=\<error|warning\>;sourcepath=\<filePath\>;linenumber=\<line\>;columnnumber=\<column\>;]\<message\>
+ *
+ * @public
+ */
+export type FileErrorFormat = 'Unix' | 'VisualStudio' | 'AzureDevOps';
+
+/**
+ * Provides options for the output message of a file error.
+ *
+ * @public
+ */
+export interface IFileErrorFormattingOptions {
+  /**
+   * The format for the error message. If no format is provided, format 'Unix' is used by default.
+   *
+   * @remarks If no format is specified and the environment variable TF_BUILD is set to 'True'
+   * (such as on Azure DevOps pipeline agents), the format 'AzureDevOps' is used by default.
+   */
+  format?: FileErrorFormat;
+
+  /**
+   * Whether or not the error is a warning. Defaults to false.
+   */
+  isWarning?: boolean;
+}
+
+const uuidFileError: string = '37a4c772-2dc8-4c66-89ae-262f8cc1f0c1';
+
+/**
+ * An `Error` subclass that should be thrown to report an unexpected state that specifically references
+ * a location in a file.
+ *
+ * @public
+ */
+export class FileError extends Error {
+  /**
+   * The path to the file that contains the error.
+   */
+  public readonly filePath: string;
+
+  /**
+   * The line number of the error in the target file.
+   */
+  public readonly line: number | undefined;
+
+  /**
+   * The column number of the error in the target file.
+   */
+  public readonly column: number | undefined;
+
+  /**
+   * Constructs a new instance of the {@link FileError} class.
+   *
+   * @param message - A message describing the error.
+   * @param filePath - The path to the file that contains the error.
+   * @param line - The line number of the error in the target file.
+   * @param column - The column number of the error in the target file.
+   */
+  public constructor(message: string, filePath: string, line?: number, column?: number) {
+    super(message);
+
+    this.filePath = Path.convertToSlashes(filePath);
```

Comment:
preserve the original  un-normalized `filePath` in a property

Reply:
`C:/folder/file.txt` is not technically a valid file path.

Reply:
Thinking about the requirements for this design:

1. The `catch` handler should be able to determine the correct absolute path that a `FileError` is referring to, without having to know what directory it might be relative to.  (Note that the correct absolute path sometimes is not well-defined, for example if we're reporting that a file cannot be found.)
2. If we doing normalization however, the `FileError` class should not completely discard its original input, since it might have diagnostic value
3. The `FileError` constructor should not do expensive work for normalization/validation; if possible this should be done on-demand
4. The `FileError` formatting should help integrate nicely with VS Code line matching, HOWEVER this is a convenience and should not dictate its design or cause awkwardness for other use cases

Reply:
VSCode understands either of the currently used formats, but expects them to be reported as absolute paths, otherwise it doesn't know what to do with them.